File: app/src/main/java/com/example/marosu/secretchat/search/SearchAdapter.java
package com.example.marosu.secretchat.search;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.marosu.secretchat.R;
import com.example.marosu.secretchat.model.entity.User;
import com.example.marosu.secretchat.util.AvatarView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.subjects.PublishSubject;

/**
 * Created by Marius-Andrei Rosu on 8/22/2017.
 */
public final class SearchAdapter extends RecyclerView.Adapter<SearchAdapter.ViewHolder> {
    private List<User> users;
    private PublishSubject<User> clickSubject = PublishSubject.create();

    public SearchAdapter(List<User> users) {
        this.users = users;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.search_item, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final User user = users.get(position);
        holder.searchItem.setOnClickListener(view -> clickSubject.onNext(user));
        holder.searchAvatar.setUser(user);
        holder.searchName.setText(user.getFullName());
        holder.searchEmail.setText(user.getEmail());
    }

    @Override
    public int getItemCount() {
        return users.size();
    }

    public void updateUsers(List<User> users) {
        this.users = users;
        notifyDataSetChanged();
    }

    public PublishSubject<User> getClickSubject() {
        return clickSubject;
    }

    protected static final class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.search_item)
        ViewGroup searchItem;

        @BindView(R.id.search_avatar)
        AvatarView searchAvatar;

        @BindView(R.id.search_name)
        TextView searchName;

        @BindView(R.id.search_email)
        TextView searchEmail;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/example/marosu/secretchat/search/SearchAdapter.kt
package com.example.marosu.secretchat.search

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.example.marosu.secretchat.R
import com.example.marosu.secretchat.model.entity.User
import com.example.marosu.secretchat.util.AvatarView
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject

/**
 * Created by Marius-Andrei Rosu on 8/29/2017.
 */
class SearchAdapter(private var users: List<User>) : RecyclerView.Adapter<SearchAdapter.ViewHolder>() {
    val clickSubject: PublishSubject<User> = PublishSubject.create<User>()

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(users[position], clickSubject)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int)
            = ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.search_item, parent, false))

    override fun getItemCount() = users.size

    fun updateUsers(users: List<User>) {
        this.users = users
        notifyDataSetChanged()
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val searchAvatar = itemView.findViewById<AvatarView>(R.id.search_avatar)
        private val searchEmail = itemView.findViewById<TextView>(R.id.search_email)
        private val searchName = itemView.findViewById<TextView>(R.id.search_name)

        fun bind(user: User, clickSubject: Subject<User>) {
            itemView.setOnClickListener { clickSubject.onNext(user) }
            searchAvatar.setUser(user)
            searchName.text = user.fullName
            searchEmail.text = user.email
        }
    }
}
